File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys

n, c = map(int, sys.stdin.readline().split())
house = []

for i in range(n):
    house.append(int(sys.stdin.readline()))
house.sort()

def can_place(house, distance, c):
    count = 1
    last_position = house[0]
    for i in range(1, len(house)):
        if house[i] - last_position >= distance:
            count += 1
            last_position = house[i]
            if count == c:
                return True
    return False

low = 1
high = house[-1] - house[0]
max_distance = 0
while low <= high:
    mid = (low + high) // 2
    if can_place(house, mid, c):
        max_distance = max(max_distance, mid)
        low = mid + 1
    else:
        high = mid - 1

print(max_distance)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void* a, const void* b) {
    return *(int*)a - *(int*)b;
}

int can_place(int* houses, int distance, int n, int c){
    int count = 1;
    int last_position = houses[0];

    for(int i = 0; i < n; i++){
        if (houses[i] - last_position >= distance){
            count += 1;
            last_position = houses[i];
            if(count == c){
                return 1;
            }
        }
    }
    return 0;
}

int find_max_distance(int* houses, int n, int c){
    qsort(houses, n, sizeof(int), compare);

    int low = 1;
    int high = houses[n - 1] - houses[0];
    int result = 0;

    while (low <= high){
        int mid = (low + high) / 2;
        if (can_place(houses, mid,n, c)){
            result = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return result;
}

int main(void)
{
    int n, c;
    scanf("%d %d", &n, &c);

    int* houses = (int*)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++){
        scanf("%d", &houses[i]);
    }

    int result = find_max_distance(houses, n, c);
    printf("%d\n", result);

    free(houses);
    return 0;
}
